### DISCUSSION TITLE
ERROR related to using Contact Properties

### DISCUSSION THREAD

**Discussion by WinsonHsia** [Anura3D/Anura3D_OpenSource#173] [Q&A]
Dear all,

I am a graduate student from Taiwan, encountering an ERROR related to using Contact Properties: "Particle 478 from element 147 lies outside the mesh." 
I am currently using GID 16.0.6 and Anura3D 2023.

The attached image shows the initial state of my model (discussions1) in step 1 without using contact, as visualized in ParaView.
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/feea48db-ebe7-4ba8-9d68-7b8fcf28ad73)

However, in CPS_002, after modifying LOADSTEPS, setting QUASISTATIC_CONVERGENCE to 0, and HOMOGENEOUS_LOCAL_DAMPING to 0.01, with CONTACT_FORMULATION enabled, an ERROR occurs after the calculation.
[ERROR] : (time step:      1)   Particle    478 from element    147 lies outside the mesh. X, Y =  -0.4023E+245   0.6968E+245
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/dc0b85e5-98c8-4d2c-9cb8-9353e9d68f9d)

Interestingly, when I slightly push the upper horizontal slope of the model inward (discussions2), as shown in the image,
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/6851fe47-60d1-41bd-a6bb-c07402ad8808)
and use the same settings in CPS_002, the simulation runs smoothly until step 41, where the same ERROR occurs.
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/7fa63e65-7d3a-43d7-bd60-c39f76da64fa)

Furthermore, in additional tests, I observed that pushing the lower horizontal slope outward to a certain extent also results in ERROR as early as step 2. I would like to seek advice on what might be causing this issue.

Are there any recommendations or suggestions? I would be truly grateful.

Below is additional information about the configuration of models.
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/24c0cbdc-6720-45c1-9531-cc2c28cddf3e)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/96f7bcb6-ef19-45e7-8a63-9de4c0013813)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/2d3e3436-aedf-4849-b502-116fe8410fe6)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/871b298a-68da-446d-9bef-c0c9af363237)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/48f8cf7d-3af4-44ce-8b2c-55250b6664d0)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/413797b1-8735-4c3f-83a0-7dc150452862)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/154399885/014291cf-b4a2-42b0-b5f9-ef5b05eec51a)

Best regards,
Winson

**Comment by FrancescaCeccato** [comment]
Hi, one possible reason for this error is the presence of a geometrical singularity point: the corner! At the corners the normal is not defined distinctively and this may generate and error in the contact the uses the normal for correction of interface velocity. You can try to smooth the corner or manually correct the direction of the normals by using the flags in CPS:
$$APPLY_CONTACT_NORMAL_CORRECTION
1
$$NUMBER_OF_CORRECTED_NORMALS
1 (or more?)
$$NODE_NORMAL_DATA
Node ID dirx diry

**Reply by weilinlee0308** [reply]
If the nodal normal vector points to 45 degrees down to the right side, is this the way to write the command?
$$NODE_NORMAL_DATA
Node ID 1 -1

**Reply by FrancescaCeccato** [reply]
let's say that the node ID is 999, then it would be:
$$NODE_NORMAL_DATA
999 0.707 -0.707
(norm of the vector must be 1)

**Reply by LongTIme1111** [reply]
Hi
How to smooth the corner

**Reply by weilinlee0308** [reply]
Please refer to page 109 - 112 of TutorialManual_2023.pdf. 
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/156299301/d1497621-ebb8-4bb2-bb19-6f17d16f7cc6)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/156299301/37eb2803-fd5c-44d7-8d42-27a2883259f7)
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/156299301/0165d13a-2d3d-4593-9eb7-dca5fa3e3d45)

**Reply by LongTIme1111** [reply]
thank you

**Reply by lucky3sunny** [reply]
> Hi, one possible reason for this error is the presence of a geometrical singularity point: the corner! At the corners the normal is not defined distinctively and this may generate and error in the contact the uses the normal for correction of interface velocity. You can try to smooth the corner or manually correct the direction of the normals by using the flags in CPS: $$APPLY_CONTACT_NORMAL_CORRECTION 1 $$NUMBER_OF_CORRECTED_NORMALS 1 (or more?) $$NODE_NORMAL_DATA Node ID dirx diry

I encountered the same error. My model consists of two parts: the upper part is an unsaturated material, while the lower part is a dry material modelled as a linear elastic solid with a very high Young’s modulus. A contact algorithm is defined at the interface between the two materials.
To resolve the error, I tried smoothing the corner and reducing the courant number, but unfortunately, neither approach worked. To investigate whether the issue was caused by an ill-defined normal vector at the corner, I redefined the upper material as a dry material and removed the associated hydraulic boundary conditions. Under these settings, the simulation ran successfully and produced correct results.

**Reply by sunny12138-9527** [reply]
<img width="1486" height="527" alt="image" src="https://github.com/user-attachments/assets/07045095-1c05-40e9-8362-68b47961cdf1" />
My contact algorithm and you also have the same problem, how to correct the normal vector definition, how to determine the node position? This is very important to me, if it is convenient to ask you ask me how you modified it? In this normal vector definition

**Comment by LongTIme1111** [comment]
Hello, I am a graduate student from the mainland, I have the same problem as you, can I ask for your contact information (e.g. email) I would like to ask for advice on how to solve this problem
